README.md
# Flutter Collapsing Toolbar

A Flutter package to create a Romantic Collapsing Toolbar. It will be useful for your awesome app.
This widget is created to be able to standalone with any views, helps you able to add it into any complex ScrollView or ListView.

This is a present from Romantic Project. More from [Romantic Developer](https://pub.dev/publishers/romanticdeveloper.com/packages)

![Demo](./demo.gif)

### Example

```dart
import 'package:flutter/material.dart';
import 'package:flutter_collapsing_toolbar/flutter_collapsing_toolbar.dart';

void main() => runApp(MyApp());

const kSampleIcons = [
  Icons.track_changes_outlined,
  Icons.receipt_long_outlined,
  Icons.wifi_protected_setup_outlined,
  Icons.add_to_home_screen_outlined,
  Icons.account_box_outlined,
];
const kSampleIconLabels = [
  'Khuyến mãi',
  'Lịch sử',
  'Chuyển tiền',
  'Nạp tiền',
  'Tài khoản',
];

class MyApp extends StatefulWidget {
  @override
  _MyAppState createState() => _MyAppState();
}

class _MyAppState extends State<MyApp> {
  final controller = ScrollController();
  double headerOffset = 0.0;

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Example',
      theme: ThemeData(
        primarySwatch: Colors.red,
      ),
      home: Scaffold(
        body: Container(
          margin: const EdgeInsets.only(top: 24.0),
          child: Column(
            children: [
              Align(
                alignment: Alignment.topCenter,
                child: CollapsingToolbar(
                  controller: controller,
                  expandedHeight: 160,
                  collapsedHeight: 64,
                  decorationForegroundColor: Color(0xffd90000),
                  decorationBackgroundColor: Colors.white,
                  onCollapsing: (double offset) {
                    setState(() {
                      headerOffset = offset;
                    });
                  },
                  leading: Container(
                    margin: EdgeInsets.only(left: 12),
                    padding: EdgeInsets.all(4),
                    decoration: ShapeDecoration(
                      color: Colors.white,
                      shape: CircleBorder(),
                    ),
                    child: Icon(
                      Icons.person,
                      size: 24,
                      color: Colors.black38,
                    ),
                  ),
                  title: Text(
                    'Romantic Developer',
                    style: TextStyle(
                      fontSize: 16,
                      color: Colors.white,
                      fontWeight: FontWeight.w600,
                    ),
                  ),
                  actions: [
                    ElevatedButton(
                      onPressed: () {},
                      style: ButtonStyle(
                        shape: MaterialStateProperty.all(CircleBorder()),
                        backgroundColor:
                            MaterialStateProperty.all(Colors.transparent),
                        elevation: MaterialStateProperty.all(0.0),
                      ),
                      child: Padding(
                        padding: const EdgeInsets.all(8.0),
                        child: Icon(
                          Icons.search,
                          color: Colors.white,
                          size: 24,
                        ),
                      ),
                    ),
                  ],
                  featureCount: 5,
                  featureIconBuilder: (context, index) {
                    return Icon(
                      kSampleIcons[index],
                      size: 54,
                      color: Colors.white,
                    );
                  },
                  featureLabelBuilder: (context, index) {
                    return Text(
                      kSampleIconLabels[index],
                      textAlign: TextAlign.center,
                      maxLines: 1,
                      style: TextStyle(
                        fontSize: 12,
                        color: Colors.white,
                      ),
                    );
                  },
                  featureOnPressed: (context, index) {},
                ),
              ),
              Expanded(
                child: Container(
                  color: Colors.white,
                  child: SingleChildScrollView(
                    controller: controller,
                    child: Column(
                      children: [
                        Container(
                          height: headerOffset,
                        ),
                        Image.asset('assets/sample.jpg'),
                        Container(
                          height: 350,
                          color: Colors.white,
                        ),
                      ],
                    ),
                  ),
                ),
              ),
            ],
          ),
        ),
      ),
    );
  }
}

```

### Development environment

```
[✓] Flutter (Channel stable, 2.0.5, on macOS 11.2.3 20D91 darwin-x64, locale en-VN)
    • Flutter version 2.0.5
    • Framework revision adc687823a (11 days ago), 2021-04-16 09:40:20 -0700
    • Engine revision b09f014e96
    • Dart version 2.12.3

[✓] Android toolchain - develop for Android devices (Android SDK version 30.0.3)
    • Platform android-30, build-tools 30.0.3
    • Java binary at: /Applications/Android Studio.app/Contents/jre/jdk/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 1.8.0_242-release-1644-b3-6915495)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS
    • Xcode at /Applications/Xcode_12.app/Contents/Developer
    • Xcode 12.4, Build version 12D4e
    • CocoaPods version 1.10.1

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 4.1)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 1.8.0_242-release-1644-b3-6915495)
```

CHANGELOG.md
## [1.0.0] - 11/08/2021

* First release version.


lib/src/collapsing_toolbar.dart
import 'dart:math' as math;

import 'package:flutter/material.dart';

import 'collapsing_toolbar_decoration.dart';

const Color _kDefaultBackgroundColor = Colors.white;
const Color _kDefaultForegroundColor = Color(0xfff42222);

/// It helps you in creating a Romantic Collapsing Toolbar.
/// This widget is created to be able to standalone with any views, helps you able to add it into any complex ScrollView or ListView.
class CollapsingToolbar extends StatefulWidget {
  final ScrollController controller;
  final double expandedHeight;
  final double collapsedHeight;
  final Function(double offset) onCollapsing;
  final Curve animationCurve;
  final Widget? leading;
  final Widget? title;
  final List<Widget>? actions;
  final Widget Function(BuildContext context, int index) featureIconBuilder;
  final Widget Function(BuildContext context, int index) featureLabelBuilder;
  final int featureCount;
  final double horizontalPadding;
  final double verticalPadding;
  final ButtonStyle? featureButtonStyle;
  final double featureCollapsedFactor;
  final Function(BuildContext context, int index) featureOnPressed;
  final Color decorationBackgroundColor;
  final Color decorationForegroundColor;
  final bool useDefaultDecoration;

  /// Create a CollapsingToolbar.
  ///
  /// [controller] is the [ScrollController] of your current scroll view, it's used to listen the scrolling offset and calculate the animation value.
  /// [collapsedHeight] & [expandedHeight] are required to handle the animation.
  /// [onCollapsing] is a callback function, it is important because it is a part that help this widget can standalone with the scroll view. During scrolling, you need to use its offset value to make a padding for your scroll|list view. For more details please see the example.
  /// [animationCurve] it helps the collapsing effect more excited.
  /// There is an [AppBar] in this. It helps easier to handle some regular features such as: a [leading] action, list of right [actions] and the [title] of the screen.
  ///
  /// [featureIconBuilder] & [featureLabelBuilder] & [featureOnPressed] & [featureButtonStyle], they help you can create your own styled feature buttons easier. [featureButtonStyle] is used to create the pressing effect.
  /// [featureCount] is count of your feature buttons. Tips: many regular designs use only 4 feature buttons.
  /// [featureCollapsedFactor] is the scale of the [Row] of feature buttons when it is collapsed.
  ///
  /// [horizontalPadding] & [verticalPadding] are used to set padding and margin the whole widget.
  ///
  /// [useDefaultDecoration] : if you want not to use the default [CollapsingToolbarDecoration], set it to false, default is true.
  ///
  /// [decorationBackgroundColor] & [decorationForegroundColor] are used to draw the default [CollapsingToolbarDecoration].
  const CollapsingToolbar({
    Key? key,
    required this.controller,
    required this.expandedHeight,
    required this.collapsedHeight,
    required this.onCollapsing,
    this.animationCurve: Curves.easeIn,
    this.leading,
    this.title,
    this.actions,
    required this.featureIconBuilder,
    required this.featureLabelBuilder,
    required this.featureCount,
    this.featureButtonStyle,
    this.featureCollapsedFactor: 0.6,
    required this.featureOnPressed,
    this.horizontalPadding: 12,
    this.verticalPadding: 12,
    this.decorationBackgroundColor: _kDefaultBackgroundColor,
    this.decorationForegroundColor: _kDefaultForegroundColor,
    this.useDefaultDecoration: true,
  }) : super(key: key);

  @override
  _CollapsingToolbarState createState() => _CollapsingToolbarState();
}

class _CollapsingToolbarState extends State<CollapsingToolbar> {
  late double _toolbarHeight;
  double _animationValue = 1.0;

  @override
  void initState() {
    _toolbarHeight = widget.expandedHeight;
    widget.controller.addListener(this._onScrolling);
    super.initState();
  }

  @override
  void dispose() {
    widget.controller.removeListener(this._onScrolling);
    super.dispose();
  }

  void _onScrolling() {
    setState(() {
      var offset = widget.controller.offset;
      var max = widget.expandedHeight;
      var min = widget.collapsedHeight;
      var gap = max - min;
      var factor = math.min(1.0, math.max(0.0, offset / gap));
      _animationValue = 1 - widget.animationCurve.transform(factor);
      _toolbarHeight = max - factor * gap;
      widget.onCollapsing(factor * gap);
    });
  }

  @override
  Widget build(BuildContext context) {
    var body = Container(
      height: _toolbarHeight,
      child: Stack(
        children: [
          Container(
            height: widget.collapsedHeight,
            child: AppBar(
              primary: false,
              toolbarHeight: widget.collapsedHeight,
              backgroundColor: Colors.transparent,
              centerTitle: false,
              titleSpacing: widget.horizontalPadding,
              elevation: 0.0,
              leading: widget.leading,
              title: Opacity(
                opacity: _animationValue,
                child: Transform.scale(
                  child: widget.title,
                  alignment: Alignment.centerLeft,
                  scale: _animationValue,
                ),
              ),
              actions: widget.actions,
            ),
          ),
          Align(
            alignment: Tween<Alignment>(
                    end: Alignment.bottomCenter, begin: Alignment.topLeft)
                .transform(_animationValue),
            child: Padding(
              padding: EdgeInsetsTween(
                  begin: EdgeInsets.only(
                      left: widget.collapsedHeight,
                      top: widget.verticalPadding),
                  end: EdgeInsets.only(
                    bottom: widget.verticalPadding,
                    left: widget.horizontalPadding,
                    right: widget.horizontalPadding,
                  )).transform(_animationValue),
              child: Transform.scale(
                alignment: Alignment.centerLeft,
                scale: Tween(begin: widget.featureCollapsedFactor, end: 1.0)
                    .transform(_animationValue),
                child: Row(
                  children: List.generate(widget.featureCount, (index) {
                    return Expanded(
                      child: ElevatedButton(
                        onPressed: () {
                          widget.featureOnPressed(context, index);
                        },
                        child: FittedBox(
                          child: Column(
                            mainAxisSize: MainAxisSize.min,
                            children: [
                              widget.featureIconBuilder(context, index),
                              Opacity(
                                opacity: _animationValue,
                                child: Transform.scale(
                                  alignment: Alignment.topCenter,
                                  scale: _animationValue,
                                  child: Padding(
                                    padding: EdgeInsets.only(
                                        top: widget.verticalPadding * 0.7),
                                    child: widget.featureLabelBuilder(
                                        context, index),
                                  ),
                                ),
                              ),
                            ],
                          ),
                        ),
                        style: widget.featureButtonStyle ??
                            ButtonStyle(
                              shape: MaterialStateProperty.all(CircleBorder()),
                              backgroundColor:
                                  MaterialStateProperty.all(Colors.transparent),
                              elevation: MaterialStateProperty.all(0.0),
                              padding:
                                  MaterialStateProperty.all(EdgeInsets.zero),
                            ),
                      ),
                    );
                  }),
                ),
              ),
            ),
          ),
        ],
      ),
    );
    return widget.useDefaultDecoration
        ? CollapsingToolbarDecoration(
            animationValue: _animationValue,
            backgroundColor: widget.decorationBackgroundColor,
            foregroundColor: widget.decorationForegroundColor,
            child: body,
          )
        : body;
  }
}


lib/src/collapsing_toolbar_decoration.dart
import 'package:flutter/material.dart';

import 'common.dart';

const double _kDecorationElevation = 8.0;

/// This is the default decoration for [CollapsingToolbar].
/// It is the default design, and you can create your own decoration that based on this solution. It's easy to handle.
class CollapsingToolbarDecoration extends StatelessWidget {
  final Widget child;
  final Color backgroundColor;
  final Color foregroundColor;
  final double animationValue;

  /// Create a default decoration.
  ///
  /// [child] is your [CollapsingToolbar].
  /// It draws some overlay ellipses with [foregroundColor] on a background with [backgroundColor].
  /// [animationValue] is calculated by the scrolling offset, see also in [CollapsingToolbar].
  const CollapsingToolbarDecoration({
    Key? key,
    required this.child,
    required this.backgroundColor,
    required this.foregroundColor,
    required this.animationValue,
  }) : super(key: key);

  @override
  Widget build(BuildContext context) {
    return CustomPaint(
      child: Padding(
        padding:
            EdgeInsets.only(bottom: animationValue * _kDecorationElevation * 3),
        child: child,
      ),
      painter: _DefaultCollapsingToolbarDecorationPainter(
        backgroundColor: backgroundColor,
        foregroundColor: foregroundColor,
        animationValue: animationValue,
        elevation: _kDecorationElevation,
      ),
    );
  }
}

class _DefaultCollapsingToolbarDecorationPainter extends CustomPainter {
  final Color backgroundColor;
  final Color foregroundColor;
  final double animationValue;
  final double elevation;

  const _DefaultCollapsingToolbarDecorationPainter({
    required this.backgroundColor,
    required this.foregroundColor,
    required this.animationValue,
    required this.elevation,
  });

  @override
  void paint(Canvas canvas, Size size) {
    final ovalH = size.height * 2;
    final ovalW = size.width * 2;
    final ellipseX = ovalW * animationValue;
    final ellipseY = ovalW * 0.5 * animationValue;
    final colorLighter = foregroundColor.lighter(0.85);
    final colorDarker = foregroundColor.darker(0.1 * animationValue);
    final colorDarker1 = foregroundColor.darker(0.15 * animationValue);

    canvas.drawRect(Rect.largest, Paint()..color = backgroundColor);

    canvas.drawPath(
        Path()
          ..addRRect(RRect.fromRectAndCorners(
            Rect.fromLTWH(ovalW * -0.25, ovalH * 0.5 * -1, ovalW, ovalH),
            bottomLeft: Radius.elliptical(ellipseX, ellipseY),
            bottomRight: Radius.elliptical(ellipseX, ellipseY),
          )),
        Paint()..color = colorLighter);

    canvas.drawPath(
        Path()
          ..addRRect(RRect.fromRectAndCorners(
            Rect.fromLTWH(ovalW * -0.25,
                ovalH * 0.5 * -1 - elevation * animationValue, ovalW, ovalH),
            bottomLeft: Radius.elliptical(ellipseX, ellipseY),
            bottomRight: Radius.elliptical(ellipseX, ellipseY),
          )),
        Paint()..color = foregroundColor);

    canvas.drawPath(
        Path()
          ..addRRect(RRect.fromRectAndCorners(
            Rect.fromLTWH(ovalW * -0.25 + (-ovalW * 0.25) * animationValue,
                ovalH * 0.5 * -1 - elevation - ovalH * 0.15, ovalW, ovalH),
            bottomRight: Radius.elliptical(ellipseY, ellipseY * 0.5),
          )),
        Paint()..color = colorDarker);

    canvas.drawPath(
        Path()
          ..addRRect(RRect.fromRectAndCorners(
            Rect.fromLTWH(ovalW * -0.25 + (-ovalW * 0.35) * animationValue,
                ovalH * 0.5 * -1 - elevation - ovalH * 0.2, ovalW, ovalH),
            bottomRight: Radius.elliptical(ellipseY, ellipseY * 0.5),
          )),
        Paint()..color = colorDarker1);
  }

  @override
  bool shouldRepaint(covariant CustomPainter oldDelegate) {
    return false;
  }
}


example/lib/main.dart
import 'package:flutter/material.dart';
import 'package:flutter_collapsing_toolbar/flutter_collapsing_toolbar.dart';

void main() => runApp(MyApp());

const kSampleIcons = [
  Icons.track_changes_outlined,
  Icons.receipt_long_outlined,
  Icons.wifi_protected_setup_outlined,
  Icons.add_to_home_screen_outlined,
  Icons.account_box_outlined,
];
const kSampleIconLabels = [
  'Khuyến mãi',
  'Lịch sử',
  'Chuyển tiền',
  'Nạp tiền',
  'Tài khoản',
];

class MyApp extends StatefulWidget {
  @override
  _MyAppState createState() => _MyAppState();
}

class _MyAppState extends State<MyApp> {
  final controller = ScrollController();
  double headerOffset = 0.0;

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Example',
      theme: ThemeData(
        primarySwatch: Colors.red,
      ),
      home: Scaffold(
        body: Container(
          margin: const EdgeInsets.only(top: 24.0),
          child: Column(
            children: [
              Align(
                alignment: Alignment.topCenter,
                child: CollapsingToolbar(
                  controller: controller,
                  expandedHeight: 160,
                  collapsedHeight: 64,
                  decorationForegroundColor: Color(0xffd90000),
                  decorationBackgroundColor: Colors.white,
                  onCollapsing: (double offset) {
                    setState(() {
                      headerOffset = offset;
                    });
                  },
                  leading: Container(
                    margin: EdgeInsets.only(left: 12),
                    padding: EdgeInsets.all(4),
                    decoration: ShapeDecoration(
                      color: Colors.white,
                      shape: CircleBorder(),
                    ),
                    child: Icon(
                      Icons.person,
                      size: 24,
                      color: Colors.black38,
                    ),
                  ),
                  title: Text(
                    'Romantic Developer',
                    style: TextStyle(
                      fontSize: 16,
                      color: Colors.white,
                      fontWeight: FontWeight.w600,
                    ),
                  ),
                  actions: [
                    ElevatedButton(
                      onPressed: () {},
                      style: ButtonStyle(
                        shape: MaterialStateProperty.all(CircleBorder()),
                        backgroundColor:
                            MaterialStateProperty.all(Colors.transparent),
                        elevation: MaterialStateProperty.all(0.0),
                      ),
                      child: Padding(
                        padding: const EdgeInsets.all(8.0),
                        child: Icon(
                          Icons.search,
                          color: Colors.white,
                          size: 24,
                        ),
                      ),
                    ),
                  ],
                  featureCount: 5,
                  featureIconBuilder: (context, index) {
                    return Icon(
                      kSampleIcons[index],
                      size: 54,
                      color: Colors.white,
                    );
                  },
                  featureLabelBuilder: (context, index) {
                    return Text(
                      kSampleIconLabels[index],
                      textAlign: TextAlign.center,
                      maxLines: 1,
                      style: TextStyle(
                        fontSize: 12,
                        color: Colors.white,
                      ),
                    );
                  },
                  featureOnPressed: (context, index) {},
                ),
              ),
              Expanded(
                child: Container(
                  color: Colors.white,
                  child: SingleChildScrollView(
                    controller: controller,
                    child: Column(
                      children: [
                        Container(
                          height: headerOffset,
                        ),
                        Image.asset('assets/sample.jpg'),
                        Container(
                          height: 350,
                          color: Colors.white,
                        ),
                      ],
                    ),
                  ),
                ),
              ),
            ],
          ),
        ),
      ),
    );
  }
}
